processor/normalizesumsprocessor/processor.go

```diff
@@ -0,0 +1,369 @@
+
+	return errors
+}
+
+//func findMetric(name string, rms pdata.ResourceMetrics) (*pdata.Metric, *pdata.MetricSlice) {
+func (nsp *NormalizeSumsProcessor) transformMetric(transform SumMetrics, resource pdata.Resource, metric *pdata.Metric, slice *pdata.MetricSlice) error {
+	switch t := metric.DataType(); t {
+	case pdata.MetricDataTypeIntSum:
+		dataPointsRemaining, processingErr := nsp.processIntSumMetric(slice, transform, resource, metric)
+		if processingErr != nil {
+			return processingErr
+		} else {
+			cleanUpMetricSlice(transform, metric, slice, dataPointsRemaining)
+		}
+	case pdata.MetricDataTypeDoubleSum:
+		dataPointsRemaining, processingErr := nsp.processDoubleSumMetric(slice, transform, resource, metric)
+		if processingErr != nil {
+			return processingErr
+		} else {
+			cleanUpMetricSlice(transform, metric, slice, dataPointsRemaining)
+		}
+	default:
+		return fmt.Errorf("Data Type not supported %s", metric.DataType())
+	}
+
+	return nil
+}
+
+func cleanUpMetricSlice(transform SumMetrics, metric *pdata.Metric, slice *pdata.MetricSlice, dataPointsRemaining int) {
+	// If there is meaningful data to send and we are renaming the metric,
+	// add the already renamed metric to the slice
+	if dataPointsRemaining > 0 && transform.NewName != "" {
+		newMetric := pdata.NewMetric()
+		metric.CopyTo(newMetric)
+		newMetric.SetName(transform.NewName)
+
+		slice.Append(*&newMetric)
+	}
+
+	// If there are no remaining data points after removing restart/start
+	// points, or this metric was renamed, remove this metric from the slice
+	if dataPointsRemaining == 0 || transform.NewName != "" {
+		metricSliceRemoveElement(slice, transform.MetricName)
+	}
+}
+
+func (nsp *NormalizeSumsProcessor) processIntSumMetric(slice *pdata.MetricSlice, transform SumMetrics, resource pdata.Resource, metric *pdata.Metric) (int, error) {
+	dps := metric.IntSum().DataPoints()
+	for i := 0; i < dps.Len(); {
+		reportData := nsp.processIntSumDataPoint(dps.At(i), resource, metric)
+
+		if reportData == false {
+			intRemoveAt(&dps, i)
+			continue
+		}
+		i++
+	}
+
+	return dps.Len(), nil
+}
+
+func (nsp *NormalizeSumsProcessor) processIntSumDataPoint(dp pdata.IntDataPoint, resource pdata.Resource, metric *pdata.Metric) bool {
+	metricIdentifier := dataPointIdentifier(resource, metric, dp.LabelsMap())
+
+	start := nsp.history[metricIdentifier]
+	// If this is the first time we've observed this unique metric,
+	// record it as the start point and do not report this data point
+	if start == nil {
+		dps := metric.IntSum().DataPoints()
+		newDP := pdata.NewIntDataPoint()
+		dps.At(0).CopyTo(newDP)
+
+		newStart := startPoint{
+			dataType:     pdata.MetricDataTypeIntSum,
+			intDataPoint: &newDP,
+			lastIntValue: newDP.Value(),
+		}
+		nsp.history[metricIdentifier] = &newStart
+
+		return false
+	}
+
+	// If this data is older than the start point, we can't
+	// meaningfully report this point
+	if dp.Timestamp() <= start.intDataPoint.Timestamp() {
+		return false
```

Comment:
I'm not sure this is the right way to always handle this, e.g. in the event of a clock reset or other oddity.

Can you (at a minimum) add a TODO here to look into other mechanisms to handle time-rewinds and (at a maximum) find a way to remember we saw a time rewind, so if we see it a second time, we treat it like a roll-over/restart ?

If you want to see some other bugs of similar fashion, we had this one in OpenCensus: https://github.com/census-instrumentation/opencensus-java/issues/2068

Reply:
Went with the TODO and log. It wouldn't be very difficult to try to track the time-rewind, but adds some complexity and increases the size of the struct being tracked for each relevant metric. If you think that makes more sense, I'm happy to toss it in.

Reply:
(a) the log line looks awesome. We might be able to create an alert if we see it often enough.
(b) My concern is on catastrophic failures (e.g. a time reset).   Would this receiver ever recover to a working state?  I think effectively it would fail to grab metrics and continue to do so.   That might be ok, but if there's something we can do to limp along reporting metrics, that's more ideal.

I think this is fine for now, because MOST failure scenarios I can imagine of this sort will either resolve or fail when actually sending to cloud monitoring due to the timestamp being wildly inaccurate.